[API] Add `guid` field to Event responses, add `/accounts/{address}/events/{creation_number}` endpoint

banool opened the pull request:
## Description
This PR addresses https://github.com/aptos-labs/aptos-core/issues/3891. In short, we're trying to remove event key from the public facing API, both for making queries and receiving responses. This PR renames it from key to guid, to make it match the data returned by the Move layer.

Note: This does not alter the `guid` field in the event data returned by move, for example in something like this:
```
  {
    "type": "0x1::aptos_governance::GovernanceEvents",
    "data": {
      "create_proposal_events": {
        "counter": "0",
        "guid": {
          "id": {
            "addr": "0x1",
            "creation_num": "9"
          }
        }
      },
      "update_config_events": {
        "counter": "0",
        "guid": {
          "id": {
            "addr": "0x1",
            "creation_num": "10"
          }
        }
      },
      "vote_events": {
        "counter": "0",
        "guid": {
          "id": {
            "addr": "0x1",
            "creation_num": "11"
          }
        }
      }
    }
  }
```
The fact that the data is one layer deeper under the `id` field doesn't seem like a big deal to me, but I just want to call it out.

Generating everything:
```
cargo run -p aptos-openapi-spec-generator -- -f yaml -o api/doc/spec.yaml && cargo run -p aptos-openapi-spec-generator -- -f json -o api/doc/spec.json
cd ecosystem/typescript/sdk
yarn generate-client
```

## Test Plan
### Manual testing
First, run a local testnet and seed it with data by running the TS SDK tests.

From this command:
```
curl http://127.0.0.1:8080/v1/transactions?limit=1000  | jq .[0].events[0]
```

Previously you'd get this:
```
{
  "key": "0x020000000000000085d74e71df00a25bdd57c7d7048324dabbe3247351ee10f3ad3de1415d721da7",
  "sequence_number": "850948",
  "type": "0x1::coin::DepositEvent",
  "data": {
    "amount": "500000"
  }
}
```

Now you get this (these are different objects naturally, I'm just demonstrating the output format):
```
{
  "guid": {
    "creation_number": "0",
    "account_address": "0xa550c18"
  },
  "sequence_number": "0",
  "type": "0x1::account::CoinRegisterEvent",
  "data": {
    "type_info": {
      "account_address": "0x1",
      "module_name": "0x6170746f735f636f696e",
      "struct_name": "0x4170746f73436f696e"
    }
  }
}
```

Previously you would have to do this to read a specific event by GUID.

Command:
```
curl https://fullnode.devnet.aptoslabs.com/v1/events/0x02000000000000000000000000000000000000000000000000000000000000000000000000000001 | jq .[0]
```

Output:
```
{
  "version": "72392853",
  "key": "0x02000000000000000000000000000000000000000000000000000000000000000000000000000001",
  "sequence_number": "145",
  "type": "0x1::reconfiguration::NewEpochEvent",
  "data": {
    "epoch": "146"
  }
}
```

Now you can do this:
```
curl http://127.0.0.1:8080/v1/accounts/0x1/events/2 | jq .[0]
```

Output (again, different object because different target nodes):
```
{
  "version": "0",
  "guid": {
    "creation_number": "2",
    "account_address": "0x1"
  },
  "sequence_number": "0",
  "type": "0x1::reconfiguration::NewEpochEvent",
  "data": {
    "epoch": "1"
  }
}
```

### Testing the TS SDK
```
yarn test
```

### Testing the Rust REST client
The Rust REST client didn't have a function for the old endpoint, so nothing changes there.

### Testing the Python SDK
The Python SDK doesn't offer event functions either.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/aptos-labs/aptos-core/4012)
<!-- Reviewable:end -->

davidiw reviewed (commented):
worth considering, let's discuss.

banool commented on api/doc/CHANGELOG.md:
```diff
@@ -4,6 +4,39 @@ All notable changes to the Aptos Node API will be captured in this file. This ch
 **Note**: The Aptos Node API does not follow semantic version while we are in active development. Instead, breaking changes will be announced with each devnet cut. Once we launch our mainnet, the API will follow semantic versioning closely.
 
+## 1.1.0 (2022-09-08)
+- The `/events/{event_key}` endpoint has been removed. In addition to it no longer featuring in any request to the API, the canonical string representation of `EventKey` no longer appears in any response from the API either. Where previously you might see a representation of an event like `OLD` below, it now looks like `NEW`. See [#4012](https://github.com/aptos-labs/aptos-core/pull/4012).
+- A new endpoint has been added for getting events by GUID: `/accounts/{address}/events/{creation_number}`. This should be more easily usable by hand than the previous API.
```
Given this is a new endpoint, this would amount to just keeping the old endpoint around for a couple of weeks. Under normal circumstances this is what I'd do, and we'd slowly deprecate the old endpoint, notify people lots, increasingly start throwing errors from the old endpoint, etc. So far we haven't really done that with the API though in the interest of moving fast. I'm happy with either approach though. Thoughts @gregnazario?

banool replied:
Either way, for the SDK I'd just delete the old function and add the new one in the next version. Users can update to the new version when they're ready, vs the API in which they have no choice.

jjleng replied:
`next version` == next major version?

gregnazario replied:
We should do a migration translation for them.

banool commented on api/doc/CHANGELOG.md:
````diff
@@ -4,6 +4,39 @@ All notable changes to the Aptos Node API will be captured in this file. This ch
+```
+{
+  "version": "0",
+  "guid": {
+    "creation_number": "2",
+    "account_address": "0x1"
+  },
````
I feel like it doesn't matter, it's pretty obvious what it is with or without the nesting. Not super tied to it either way though.

banool commented:
Seems like this indexer test (`indexer::tailer::test::test_parsing_and_writing`) is failing in CI but I can't repro it locally. Trying to fix it anyway. On the topic, I think defining a transaction as JSON and parsing it here provides very little value to the test, it should just define one in code directly.

gregnazario reviewed (commented):
Okay there are two things we should do here:
* Keep the events API, we can deprecate it
* Keep both the key and the GUID, and we can deprecate the key

banool commented:
We're very much running out of time for this, if we want to do deprecate this way (land new endpoint next week, mark old one as deprecated, remove old endpoint the week after) we won't get manage to fully remove the old endpoint until the 22nd. If that's okay despite the freeze, then sure we can do that.

banool commented:
Addressed feedback:
- Kept old endpoint for now.
- Updated changelog to say that the old endpoint is deprecated and will be removed in the next release.
- Included a migration translation in both changelogs.

I verified that the old endpoint still works for now, though the response uses the new guid:
```
$ curl 127.0.0.1:8080/v1/events/0x02000000000000000000000000000000000000000000000000000000000000000000000000000001 | jq .[0]
{
  "version": "0",
  "guid": {
    "creation_number": "2",
    "account_address": "0x1"
  },
  "sequence_number": "0",
  "type": "0x1::reconfiguration::NewEpochEvent",
  "data": {
    "epoch": "1"
  }
}
```

banool commented:
Once this is all stamped, I'll land it prior to next devnet cut, and then land another PR after that cut to remove the old endpoint and related type entirely after that devnet cut, but prior to the one after that.

gregnazario commented on api/types/src/transaction.rs:
```diff
@@ -475,7 +475,10 @@ pub struct BlockMetadataTransaction {
 /// An event from a transaction
 #[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, Object)]
 pub struct Event {
-    pub key: EventKey,
```
Can you keep, then remove this field for the deprecation path?

gregnazario commented on api/goldens/aptos_api__tests__events_test__test_get_events.json:
```diff
@@ -1 +1,18 @@
+    "data": {
+      "type_info": {
+        "account_address": "0x1",
+        "module_name": "0x6170746f735f636f696e",
+        "struct_name": "0x4170746f73436f696e"
```
This seems wrong?  Oh, it's BCS encoded

gregnazario reviewed (approved):
Just add the key back in (even just as a `Option<String>` or `String`), and we can remove it in the upcoming update

davidiw commented on api/doc/CHANGELOG.md:
```diff
@@ -4,6 +4,50 @@ All notable changes to the Aptos Node API will be captured in this file. This ch
 
+## 1.1.0 (2022-09-08)
+- A new endpoint has been added for getting events by GUID: `/accounts/{address}/events/{creation_number}`. If you would make a request to `/events/{event_key}` like in `Example A` below, you would use the new endpoint like in `Example B`.
+- **[Deprecated]** The `/events/{event_key}` endpoint is now deprecated. In the next release it will be removed entirely. You must migrate to the new endpoint, `/accounts/{address}/events/{creation_number}`, by then.
```
We should probably give a 2 week cycle for deprecation / deletion now.

davidiw commented on api/doc/CHANGELOG.md:
````diff
@@ -4,6 +4,50 @@ All notable changes to the Aptos Node API will be captured in this file. This ch
+}
+```
+
+Output B (new representation):
+```
+{
+  "version": "0",
+  "guid": {
````
can we keep key here temporarily?

banool replied:
Will do tomorrow 👍

davidiw reviewed (commented):
I think we possibly want to keep both key formats for a little while. This will also affect the transactions api.

@kent-white does this affect you?
@MartianAbhishek you?
@borispovod you?
@CapCap do you know anyone else to poke if we just want to rip this bandaid off?

banool commented:
Sure I'll add back the key field in the old format alongside the new guid field, then remove it later.

jjleng commented on ecosystem/typescript/sdk/src/aptos_client.ts:
```diff
@@ -267,9 +269,31 @@ export class AptosClient {
+    return this.client.events.getEventsByCreationNumber(HexString.ensure(address).hex(), creationNumber.toString());
+  }
+
+  /**
```
add `@deprecated` and tell people start using getEventsByCreationNumber

banool replied:
This is not deprecated, `getEventsByEventKey` is. See above where I've added `@deprecated` to that.

jjleng commented on ecosystem/typescript/sdk/src/aptos_client.ts:
```diff
@@ -267,9 +269,31 @@ export class AptosClient {
+  @parseApiError
+  async getEventsByCreationNumber(
+    address: MaybeHexString,
+    creationNumber: number | BigInt | string,
```
nit: BigInt is a wrapper type. Let's use `bigint` here. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt#type_information
